add copyright and authors to json report

notofug opened the pull request:
This PR adds 'copyright' and 'authors' to the json output for use in post-processing to generate complete list of third-party components. I hope this is in line with your plan for the tool

Another thing I would like in the json is a 'pointer' to the associated downloaded license.html but that is not included here.

![image](https://github.com/sensslen/nuget-license/assets/40484721/84a5f1c1-2d80-433b-9077-59661f40ebc2)

sensslen reviewed (changes requested):
Unfortunately I cannot let this go in as is.
1. While the tests pass now, the functionality is not tested
2. The tabular output does not provide the new information

notofug commented:
Latest commits addresses your comments.

But are you sure you want all properties in text-output also, it does not necessarily scale very well since it potentially make a very 'wide' output ;
![image](https://github.com/sensslen/nuget-license/assets/40484721/20a0def3-174d-4050-8674-700bfff18771)

sensslen commented:
@notofug I'm fully aware of this and have been debating whether it's worth to add a command line flag that allows to enable/disable this. In general I'm not at all a fan of too many flags. I'm also not sure whether the tabular output should be moving towards a markdown compatible output which in turn will allow for the table to be rendered as html and thus avoiding the issues you describe.

sensslen commented on src/NuGetUtility/Wrapper/NuGetWrapper/Protocol/WrappedPackageMetadata.cs:
```diff
@@ -28,6 +28,10 @@ public WrappedPackageMetadata(ManifestMetadata metadata)
 
+        public string Copyright => _metadata.Copyright;
+
+        public string Authors => string.Join(',', _metadata.Authors); // https://learn.microsoft.com/en-us/nuget/reference/nuspec#authors
```
Can you please keep the authors as array - it will be much easier to parse the output by a machine if we concatinate with comma.

sensslen commented on tests/NuGetUtility.Test/LicenseValidator/LicenseValidatorTest.cs:
```diff
@@ -668,6 +722,66 @@ private IPackageMetadata SetupPackage(string packageId, INuGetVersion packageVer
+                            packageVersion,
+                            _projectUrl.ToString(),
+                            null,
+                            copyrigth,
```
```suggestion
                            copyright,
```

sensslen commented on tests/NuGetUtility.Test/LicenseValidator/LicenseValidatorTest.cs:
```diff
@@ -668,6 +722,66 @@ private IPackageMetadata SetupPackage(string packageId, INuGetVersion packageVer
+                            LicenseInformationOrigin.Ignored)
+                    })
+                    .Using(new LicenseValidationResultValueEqualityComparer()));
+        }
```
I'm missing a test with multiple authors here.... With AutoData you should be able to easily create an array of strings

sensslen commented on tests/NuGetUtility.Test/Output/TestBase.cs:
```diff
@@ -21,13 +21,17 @@ public void SetUp()
                         new NuGetVersion(f.System.Semver()),
                         f.Internet.Url(),
                         f.Hacker.Phrase(),
+                        null,
```
Can you please add tests for the filed formatting here as well?

sensslen replied:
I'll take a look when I get time....

sensslen commented on src/NuGetUtility/LicenseValidator/LicenseValidationResult.cs:
```diff
@@ -7,7 +7,7 @@ public record LicenseValidationResult(string PackageId,
         string? License,
         string? Copyright,
-        string? Authors,
+        IEnumerable<string>? Authors,
```
can you please make the enumerable non nullable?

notofug replied:
Oh crap, that alters the json-output to always have a field there even when empty (and breaks some tests), maybe revert it ?

sensslen replied:
I'm not too concerned about the tests breaking, as this is expected when adding new fields. We could change the serialization options to not output empty enumerables... However I did not see that the authors field was already comma seperated in the cached data. I think in that context keeping them as comma separated string is a good option.

notofug replied:
There are 2 places from cached data ;
NuGet.Packaging.ManifestMetadata.Authors (IEnumerable<string>)
NuGet.Protocol.Core.Types.IPackageSearchMetadata.Authors (string)
Either way we need to transform one of them, I initially picked plain string since it is a little easier to handle

sensslen replied:
Let's go with string then....

sensslen commented:
@notofug would you mind putting in some documentation on what is supported for which output variant?

notofug commented:
All output-variants are extended with 'Copyright' and 'Authors'. Should that be mentioned in the readme, or is it something else you have in mind?
Perhaps the readme should mention that the tool also processes packages.config and vcxproj's ?

sensslen commented:
Oh, right - I forgot that you added that. I think it's a good idea to add some documentation about c++ project support. However this belongs in a different PR...

notofug commented:
OK that set off merge problems so now it doesnt work aparently

notofug commented on src/NuGetUtility/Output/Table/TableOutputFormatter.cs:
```diff
@@ -22,6 +22,9 @@ public async Task Write(Stream stream, IList<LicenseValidationResult> results)
                 new ColumnDefinition("License Information Origin", license => license.LicenseInformationOrigin, license => true, true),
                 new ColumnDefinition("License Expression", license => license.License, license => license.License != null),
+                new ColumnDefinition("Copyright", license => license.Copyright, license => license.Copyright != null),
+                new ColumnDefinition("Copyrigth", license => license.Copyright, license => license.Copyright != null),
```
yeah, was fixed already but came in again in megamerge.
There are so many changes in this PR now that I dont really trust it to be correct or reviewable anymore so maybe better if you cherry-pick if you want to bring it into main?

sensslen replied:
Most changes are expected output... if you ignore all .txt files, things are not bad anymore

sensslen commented:
This PR has been superseeded by #50
